src/Microsoft.Diagnostics.Monitoring/DiagnosticServices.cs

```diff
@@ -21,23 +21,29 @@ public sealed class DiagnosticServices : IDiagnosticServices
     {
         private const int DockerEntrypointProcessId = 1;
 
+        private readonly IDiagnosticsConnectionsSourceInternal _connectionsSource;
         private readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
 
-        public IEnumerable<int> GetProcesses()
+        public DiagnosticServices(IDiagnosticsConnectionsSource connectionsSource)
+        {
+            _connectionsSource = (IDiagnosticsConnectionsSourceInternal)connectionsSource;
+        }
+
+        public async Task<IEnumerable<IProcessInfo>> GetProcessesAsync(CancellationToken token)
         {
             try
             {
-                //TODO This won't work properly with multi-container scenarios that don't share the process space.
-                //TODO We will need to use DiagnosticsAgent if we are the server.
-                return DiagnosticsClient.GetPublishedProcesses();
+                var connections = await _connectionsSource.GetConnectionsAsync(token);
+
+                return connections.Select(c => new ProcessInfo(c.RuntimeInstanceCookie, c.ProcessId)).ToList();
```

Comment:
nit: I don't feel strongly about this, but instead of making another copy here, consider returning the enumerable directly or perhaps making the contract IReadOnlyList?